tasks/sierpinski-triangle/src/main.rs

```diff
@@ -1,27 +1,55 @@
 use std::iter::repeat;
 
-fn main() {
-    let order = 4;
-    let height = 1 << order;
-    let mut state: Vec<bool> = repeat(true).take(height + 1).collect();
+fn sierpinski(order: usize) {
+    let mut triangle = vec!["*".to_string()];
+    for i in 0..order {
+        let space = repeat(' ').take(2_usize.pow(i as u32)).collect::<String>();
+
+        // save original state
+        let mut d = triangle.clone();
+
+        // extend existing lines
+        d.iter_mut().for_each(|r| {
+            let new_row = format!("{}{}{}", space, r, space);
+            *r = new_row;
+        });
 
-    // Compute the triangle line-by-line by viewing it as Pascal's triangle (mod 2)
-    for i in 0..height {
-        for _ in 0..height - i - 1 {
-            print!(" ");
-        }
+        // add new lines
+        triangle.iter().for_each(|r| {
+            let new_row = format!("{}{}{}", r, " ", r);
+            d.push(new_row);
+        });
+
+        triangle = d;
+    }
 
-        for filled in state.iter().take(i + 1) {
-            let fill = if *filled { "*" } else { " " };
+    triangle.iter().for_each(|r| println!("{}", r));
+}
+fn main() {
+    let order = std::env::args()
+        .nth(1)
+        .unwrap_or_else(|| "4".to_string())
+        .parse::<usize>()
+        .unwrap();
+
+    sierpinski(order);
+}
 
-            print!(" {}", fill);
-        }
+#[cfg(test)]
+mod tests {
+    use assert_cmd::Command;
 
-        // Compute the next line
-        for j in (i as i32..0).rev().step_by(1) {
-            state[j as usize] ^= state[(j - 1) as usize];
-        }
+    #[test]
+    fn test_outputs() {
+        let mut cmd = Command::cargo_bin("sierpinski-triangle").unwrap();
+        cmd.arg("2")
+            .assert()
+            .success()
+            .stdout("   *   \n  * *  \n *   * \n* * * *\n");
```

Comment:
Let's use [indoc](https://docs.rs/indoc/1.0.3/indoc/) to make these tests a bit more readable.

Reply:
That would be a great solution, however when I have tried it seems it "eats" the leftmost spaces in the first line (see examples) , so the assertion fails, as the output actually starts with spaces. Any idea how this could be avoided?

Reply:
This works for me (note the trailing whitespace):

```
#[cfg(test)]
#[rustfmt::skip]
mod tests {
    use assert_cmd::Command;
    use indoc::indoc;

    #[test]
    fn order_2() {
        let mut cmd = Command::cargo_bin("sierpinski-triangle").unwrap();
        cmd.arg("2").assert().success().stdout(indoc! {"
               *   
              * *  
             *   * 
            * * * *
        "});
    }

    #[test]
    fn order_4() {
        let mut cmd = Command::cargo_bin("sierpinski-triangle").unwrap();
        cmd.arg("4").assert().success().stdout(indoc! {"
                           *               
                          * *              
                         *   *             
                        * * * *            
                       *       *           
                      * *     * *          
                     *   *   *   *         
                    * * * * * * * *        
                   *               *       
                  * *             * *      
                 *   *           *   *     
                * * * *         * * * *    
               *       *       *       *   
              * *     * *     * *     * *  
             *   *   *   *   *   *   *   * 
            * * * * * * * * * * * * * * * *
        "});
    }
}
```

Reply:
Thanks, that was the trick!